[libc++] Eliminate extra allocations from `std::move(oss).str()`

AMP999 opened the pull request:
Add test coverage for the new behaviors, especially to verify that the returned string uses the correct allocator.
Fixes https://github.com/llvm/llvm-project/issues/64644

Migrated from https://reviews.llvm.org/D157776 — @philnik777  @pfusik 
@ldionne @mordante 
 please take another look!

philnik777 commented on libcxx/test/std/input.output/string.streams/istringstream/istringstream.members/str.allocator_propagation.pass.cpp:
```diff
@@ -0,0 +1,165 @@
+#include "test_macros.h"
+
+template <class T>
+struct SocccAllocator {
```
I think it would make sense to move this to support/allocators.h`.

philnik777 commented on libcxx/test/std/input.output/string.streams/istringstream/istringstream.members/str.allocator_propagation.pass.cpp:
```diff
@@ -0,0 +1,165 @@
+//===----------------------------------------------------------------------===//
```
IIRC the changes here are not to make us conforming, so all the tests should be either under `libcxx/test/libcxx` or guarded with `#ifdef _LIBCPP_VERSION`.

ldionne replied:
Per https://github.com/llvm/llvm-project/issues/64644#issuecomment-1748957379, it looks like this is in fact necessary for conformance.

philnik777 reviewed (changes requested):
Looks mostly good. Just some nits regarding the tests.

AMP999 commented:
@philnik @ldionne 
Re "there is no requirement in the standard," I think the use of the term "move constructed" in https://eel.is/c++draft/stringbuf#members-10.sentence-1 is that requirement. The original paper P0408 also contains this test case https://github.com/PeterSommerlad/SC22WG21_Papers/blob/master/workspace/Test_basic_stringbuf_efficient/src/Testp0407-p0408-efficientStringstreams.cpp#L105-L112 which won't succeed unless we transfer ownership here. Transfer-of-ownership is clearly P0408's intent.

AMP999 committed:
[libc++] Eliminate extra allocations from `std::move(oss).str()`

Add test coverage for the new behaviors, especially to verify that
the returned string uses the correct allocator.

Fixes https://github.com/llvm/llvm-project/issues/64644

AMP999 commented on libcxx/include/string:
```diff
@@ -1329,6 +1324,20 @@ public:
+#if _LIBCPP_STD_VER >= 20
+  _LIBCPP_HIDE_FROM_ABI constexpr
+  void __move_assign(basic_string&& __str, size_type __pos, size_type __len) {
+    // Pilfer the allocation from __str.
```
Replacement is done.

AMP999 commented:
@philnik777 can you land this for me?

philnik777 commented:
The CI didn't pass. Please update the PR and ping me again when the CI is green.

AMP999 commented:
CI has failed due to test failures. I've checked them locally on the `main` branch, and all the same tests failed on the `main` too. So, I think they are not related to my changes. Can you help me?

AMP999 commented:
Thank you! It turned out that it was actually my fault. Now it's fixed, and CI is green.
@philnik777, can you land this for me, please?
